src/ja/server/dispatcher/dispatcher.py

```diff
@@ -25,13 +27,26 @@ def set_distribution(self, job_distribution: JobDistribution) -> None:
 
         @param job_distribution the new job distribution.
         """
-
-    def _get_changed_jobs(self, jobs: List[Job]) -> List[Job]:
-        """!
-        @return list of the jobs whose state has been changed.
-        """
-
-    def _reset(self) -> None:
-        """!
-        Dispatches the added jobs, pauses the paused jobs and cancels the cancelled jobs.
-        """
+        for job_entry in job_distribution:
+            job = job_entry.job
+            work_machine = job_entry.assigned_machine
+            previous_status = self._previous_statuses.pop(job.uid, None)
+            if previous_status is None or job.status != previous_status:
+                proxy = self._proxy_factory.get_proxy(work_machine)
+                if job.status == JobStatus.RUNNING:
+                    if previous_status is None or previous_status == JobStatus.QUEUED:
+                        proxy.dispatch_job(job)
+                    elif previous_status == JobStatus.PAUSED:
+                        proxy.resume_job(job.uid)
+                    self._previous_statuses[job.uid] = job.status
+                elif job.status == JobStatus.CANCELLED:
+                    proxy.cancel_job(job.uid)  # Do not add job back to self._previous_statuses
+                elif job.status == JobStatus.PAUSED:
+                    proxy.pause_job(job.uid)
+                    self._previous_statuses[job.uid] = job.status
+                elif job.status == JobStatus.QUEUED:
+                    self._previous_statuses[job.uid] = job.status
+                elif job.status in [JobStatus.CRASHED, JobStatus.DONE]:
```

Comment:
`CRASHED` and `DONE` should never make it to the dispatcher.
The scheduler does not concern itself with those states at all.

Reply:
This also means that the dispatcher should clear its previous status for any job which is not present in the new distribution.